plugins/providers/shield/provider.go

```diff
@@ -0,0 +1,267 @@
+	return pv.GetRoles(pc, resourceType)
+}
+
+func (p *provider) GrantAccess(pc *domain.ProviderConfig, a domain.Grant) error {
+	// TODO: validate provider config and appeal
+
+	var creds Credentials
+	if err := mapstructure.Decode(pc.Credentials, &creds); err != nil {
+		return err
+	}
+	client, err := p.getClient(pc.URN, creds)
+	if err != nil {
+		return err
+	}
+
+	permissions := getPermissions(a)
+
+	var user *User
+	if user, err = client.GetSelfUser(a.AccountID); err != nil {
+		return nil
+	}
+
+	if a.Resource.Type == ResourceTypeTeam {
+		t := new(Team)
+		if err := t.FromDomain(a.Resource); err != nil {
+			return err
+		}
+		for _, p := range permissions {
+			if err := client.GrantTeamAccess(t, user.ID, string(p)); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	} else if a.Resource.Type == ResourceTypeProject {
+		pj := new(Project)
+		if err := pj.FromDomain(a.Resource); err != nil {
+			return err
+		}
+		for _, p := range permissions {
+			if err := client.GrantProjectAccess(pj, user.ID, string(p)); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	} else if a.Resource.Type == ResourceTypeOrganization {
+		o := new(Organization)
+		if err := o.FromDomain(a.Resource); err != nil {
+			return err
+		}
+		for _, p := range permissions {
+			if err := client.GrantOrganizationAccess(o, user.ID, string(p)); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	return ErrInvalidResourceType
+}
+
+func (p *provider) RevokeAccess(pc *domain.ProviderConfig, a domain.Grant) error {
+	// TODO: validate provider config and appeal
+
+	var creds Credentials
+	if err := mapstructure.Decode(pc.Credentials, &creds); err != nil {
+		return err
+	}
+	client, err := p.getClient(pc.URN, creds)
+	if err != nil {
+		return err
+	}
+
+	permissions := getPermissions(a)
+
+	var user *User
+	if user, err = client.GetSelfUser(a.AccountID); err != nil {
+		return nil
+	}
+
+	if a.Resource.Type == ResourceTypeTeam {
+		t := new(Team)
+		if err := t.FromDomain(a.Resource); err != nil {
+			return err
+		}
+		for _, p := range permissions {
+			if err := client.RevokeTeamAccess(t, user.ID, string(p)); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	} else if a.Resource.Type == ResourceTypeProject {
+		pj := new(Project)
+		if err := pj.FromDomain(a.Resource); err != nil {
+			return err
+		}
+		for _, p := range permissions {
+			if err := client.RevokeProjectAccess(pj, user.ID, string(p)); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	} else if a.Resource.Type == ResourceTypeOrganization {
+		o := new(Organization)
+		if err := o.FromDomain(a.Resource); err != nil {
+			return err
+		}
+		for _, p := range permissions {
+			if err := client.RevokeOrganizationAccess(o, user.ID, string(p)); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	return ErrInvalidResourceType
+}
+
+func getPermissions(a domain.Grant) []Permission {
```

Comment:
this seems like a generic domain-level functionality... should this be here in the shield provider code?